lucene/queries/src/test/org/apache/lucene/queries/function/TestFunctionScoreQuery.java

```diff
@@ -244,4 +247,33 @@ public void testAccessToValueSource() throws Exception {
 
   }
 
+  public void testScoreMode() throws Exception {
+    // Value Source doesn't need scores
+    assertInnerScoreMode(ScoreMode.COMPLETE_NO_SCORES, ScoreMode.COMPLETE, DoubleValuesSource.fromDoubleField("foo"));
```

Comment:
While this test makes the current code pass, I'm trying to understand why it is correct. Why do we pass a COMPLETE_NO_SCORES to the underlying weight when the function's ValueSource doesn't need scores?

Reply:
If the function's values source doesn't need scores, then Scorer#score() will never be called on the inner scorers, so we can treat them as filters, which is what `COMPLETE_NO_SCORES` does.

Reply:
Ah, right. ValueSources like `MultiplicativeBoostValuesSource` will need scores because it uses the `DoubleValues scores` argument.